match/lib/match/storage/google_cloud_storage.rb

```diff
@@ -0,0 +1,245 @@
+          env_name: "MATCH_GOOGLE_CLOUD_BUCKET_NAME",
+          description: "Name of the Google Cloud Storage bucket to use",
+          optional: true
+        )
+      )
+      Match::Options.append_option(
+        FastlaneCore::ConfigItem.new(
+          key: :google_cloud_keys_file,
+          env_name: "MATCH_GOOGLE_CLOUD_KEYS_FILE",
+          description: "Path to the `#{DEFAULT_KEYS_FILE_NAME}` file",
+          optional: true,
+          verify_block: proc do |value|
+            UI.user_error!("Could not find keys file at path '#{File.expand_path(value)}'") unless File.exist?(value)
+          end
+        )
+      )
+
+      def self.configure(params)
+        if params[:git_url].to_s.length > 0
+          UI.important("Looks like you still define a `git_url` somewhere, even though")
+          UI.important("you use Google Cloud Storage. You can remove the `git_url`")
+          UI.important("from your Matchfile and Fastfile")
+          UI.message("The above is just a warning, fastlane will continue as usual now...")
+        end
+
+        return self.new(
+          type: params[:type].to_s,
+          platform: params[:platform].to_s,
+          google_cloud_bucket_name: params[:google_cloud_bucket_name],
+          google_cloud_keys_file: params[:google_cloud_keys_file]
+        )
+      end
+
+      def initialize(type: nil,
+                     platform: nil,
+                     google_cloud_bucket_name: nil,
+                     google_cloud_keys_file: nil)
+        self.type = type if type
+        self.platform = platform if platform
+        self.google_cloud_keys_file = google_cloud_keys_file
+        self.bucket_name = google_cloud_bucket_name
+
+        ensure_keys_file_exists
+
+        # Extract the Project ID from the `JSON` file
+        # so the user doesn't have to provide it manually
+        keys_file_content = JSON.parse(File.read(self.google_cloud_keys_file))
+        project_id = keys_file_content["project_id"]
+        if project_id.to_s.length == 0
+          UI.user_error!("Provided keys file on path #{File.expand_path(self.google_cloud_keys_file)} doesn't include required value for `project_id`")
+        end
+
+        # Create the Google Cloud Storage client
+        # If the JSON auth file is invalid, this line will
+        # raise an exception
+        begin
+          self.gc_storage = Google::Cloud::Storage.new(
+            credentials: self.google_cloud_keys_file,
+            project_id: project_id
+          )
+        rescue => ex
+          UI.error(ex)
+          UI.verbose(ex.backtrace.join("\n"))
+          UI.user_error!("Couldn't log into your Google Cloud account using the provided JSON file at path '#{File.expand_path(self.google_cloud_keys_file)}'")
+        end
+
+        ensure_bucket_is_selected
+      end
+
+      def download
+        # Check if we already have a functional working_directory
+        return self.working_directory if @working_directory
```

Comment:
I am unclear if this is intentional. If we have set `@working_directory`, we do not perform a download? Perhaps there should be a warning to the user that the download was deemed unnecessary?

Also, this function does not return anything if the download is successful, but returns when the download was skipped. Is this return value used anywhere?